core/src/main/scala/org/apache/spark/rdd/UnionRDD.scala

```diff
@@ -62,8 +62,14 @@ class UnionRDD[T: ClassTag](
     var rdds: Seq[RDD[T]])
   extends RDD[T](sc, Nil) {  // Nil since we implement getDependencies
 
+  // visible for testing
+  private[spark] val isPartitionEvalParallel: Boolean =
+    rdds.length > conf.getInt("spark.rdd.parallelListingThreshold", 10)
+
   override def getPartitions: Array[Partition] = {
-    val array = new Array[Partition](rdds.map(_.partitions.length).sum)
+    val parRDDs = if (isPartitionEvalParallel) rdds.par else rdds
```

Comment:
Right, I suppose I'm saying much more than just this fails in that case. I support your suggestion. The original version always made a brand-new pool on each call; we should avoid that at least, and also not keep a bunch of threads around when not used if that's possible in ForkJoinSupport.